File: benchmarks/comprehensive_benchmark.py
#!/usr/bin/env python3
# Comprehensive benchmark for Python
# Equivalent to Sentra benchmark for fair comparison

import time

print("=== PYTHON COMPREHENSIVE BENCHMARK ===")
print("")

# Test 1: Arithmetic Operations (50k iterations)
print("Test 1: Arithmetic operations...")
start_time = time.time() * 1000  # Convert to milliseconds

sum_val = 0
for i in range(50000):
    sum_val = sum_val + i * 2 - 1
    sum_val = sum_val / 1.1
    sum_val = sum_val % 10000

arithmetic_time = time.time() * 1000 - start_time
print(f"Arithmetic (50k ops): {arithmetic_time:.0f}ms")

# Test 2: Array Operations
print("")
print("Test 2: Array operations...")
start_time = time.time() * 1000

arr = []
for i in range(5000):
    arr.append(i)
    arr.append(i * 2)

total = 0
for val in arr:
    total = total + val

array_time = time.time() * 1000 - start_time
print(f"Array (10k elements): {array_time:.0f}ms")

# Test 3: Map (Dictionary) Operations
print("")
print("Test 3: Map operations...")
start_time = time.time() * 1000

map_data = {}
for i in range(2000):
    map_data[f"key{i}"] = i * 10
    map_data[f"data{i}"] = i * 20

map_count = 0
for key in map_data:
    map_count = map_count + 1

map_time = time.time() * 1000 - start_time
print(f"Map (4k entries): {map_time:.0f}ms")

# Test 4: String Operations
print("")
print("Test 4: String operations...")
start_time = time.time() * 1000

text = "benchmark"
for i in range(1000):
    text = text + f" test {i}"
    if len(text) > 50000:
        text = "reset"

string_time = time.time() * 1000 - start_time
print(f"String (1k concatenations): {string_time:.0f}ms")

# Test 5: Function Calls
print("")
print("Test 5: Function calls...")

def calculate(a, b):
    return a * b + (a - b)

start_time = time.time() * 1000

func_result = 0
for i in range(10000):
    func_result = func_result + calculate(i, i + 1)

function_time = time.time() * 1000 - start_time
print(f"Function calls (10k): {function_time:.0f}ms")

# Test 6: Nested Loops
print("")
print("Test 6: Nested loops...")
start_time = time.time() * 1000

nested_sum = 0
for i in range(100):
    for j in range(100):
        nested_sum = nested_sum + i * j

nested_time = time.time() * 1000 - start_time
print(f"Nested loops (10k iterations): {nested_time:.0f}ms")

# Total time
total_time = arithmetic_time + array_time + map_time + string_time + function_time + nested_time
print("")
print("=== BENCHMARK COMPLETE ===")
print(f"Total execution time: {total_time:.0f}ms")
print("Platform: Python (CPython)")

File: benchmarks/comprehensive_benchmark.rb
#!/usr/bin/env ruby
# Comprehensive benchmark for Ruby
# Equivalent to Sentra benchmark for fair comparison

puts "=== RUBY COMPREHENSIVE BENCHMARK ==="
puts ""

# Test 1: Arithmetic Operations (50k iterations)
puts "Test 1: Arithmetic operations..."
start_time = Time.now.to_f * 1000

sum = 0
50000.times do |i|
  sum = sum + i * 2 - 1
  sum = sum / 1.1
  sum = sum % 10000
end

arithmetic_time = Time.now.to_f * 1000 - start_time
puts "Arithmetic (50k ops): #{arithmetic_time.round}ms"

# Test 2: Array Operations
puts ""
puts "Test 2: Array operations..."
start_time = Time.now.to_f * 1000

arr = []
5000.times do |i|
  arr.push(i)
  arr.push(i * 2)
end

total = 0
arr.each do |val|
  total = total + val
end

array_time = Time.now.to_f * 1000 - start_time
puts "Array (10k elements): #{array_time.round}ms"

# Test 3: Map (Hash) Operations
puts ""
puts "Test 3: Map operations..."
start_time = Time.now.to_f * 1000

map_data = {}
2000.times do |i|
  map_data["key#{i}"] = i * 10
  map_data["data#{i}"] = i * 20
end

map_count = 0
map_data.each_key do |key|
  map_count = map_count + 1
end

map_time = Time.now.to_f * 1000 - start_time
puts "Map (4k entries): #{map_time.round}ms"

# Test 4: String Operations
puts ""
puts "Test 4: String operations..."
start_time = Time.now.to_f * 1000

text = "benchmark"
1000.times do |i|
  text = text + " test #{i}"
  if text.length > 50000
    text = "reset"
  end
end

string_time = Time.now.to_f * 1000 - start_time
puts "String (1k concatenations): #{string_time.round}ms"

# Test 5: Function Calls
puts ""
puts "Test 5: Function calls..."

def calculate(a, b)
  a * b + (a - b)
end

start_time = Time.now.to_f * 1000

func_result = 0
10000.times do |i|
  func_result = func_result + calculate(i, i + 1)
end

function_time = Time.now.to_f * 1000 - start_time
puts "Function calls (10k): #{function_time.round}ms"

# Test 6: Nested Loops
puts ""
puts "Test 6: Nested loops..."
start_time = Time.now.to_f * 1000

nested_sum = 0
100.times do |i|
  100.times do |j|
    nested_sum = nested_sum + i * j
  end
end

nested_time = Time.now.to_f * 1000 - start_time
puts "Nested loops (10k iterations): #{nested_time.round}ms"

# Total time
total_time = arithmetic_time + array_time + map_time + string_time + function_time + nested_time
puts ""
puts "=== BENCHMARK COMPLETE ==="
puts "Total execution time: #{total_time.round}ms"
puts "Platform: Ruby (YARV)"